### PR TITLE
Add completions to server

### PR DESCRIPTION
**Pull request by beyang**
Moves completions from extension -> Anthropic to extension -> Sourcegraph -> Anthropic, so we can use the Anthropic access token from the Sourcegraph instance.

* Adds the `.api/completions/code` endpoint (to accompany the existing `.api/completions/stream` endpoint for chat)
* Updates the VS Code extension to use this endpoint

## Test plan

Affects only Cody (experimental)

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.32 kb) | 0.00% (+0.32 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/bl/completions-server/compare-report.html;tab=live_object) for a full comparison between the commits 34dde4a and 3622820 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts [hunk 1]
```diff
@@ -5,9 +5,58 @@ import { isError } from '../../utils'
...
+    public async complete(params: CodeCompletionParameters): Promise<CodeCompletionResponse> {
+        const requestFn = this.codeCompletionsEndpoint.startsWith('https://') ? https.request : http.request
+        const completion = await new Promise<CodeCompletionResponse>((resolve, reject) => {
+            const req = requestFn(
+                this.codeCompletionsEndpoint,
+                {
+                    method: 'POST',
+                    headers: {
+                        'Content-Type': 'application/json',
+                        ...(this.accessToken ? { Authorization: `token ${this.accessToken}` } : null),
```
Comment (dominiccooney) [inline client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts:20]: First time I've seen this pattern of spreading a ternary which could be null. Neat.

File: client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts [hunk 2]
```diff
@@ -5,9 +5,58 @@ import { isError } from '../../utils'
...
+                    },
+                    // So we can send requests to the Sourcegraph local development instance, which has an incompatible cert.
+                    rejectUnauthorized: this.mode === 'production',
```
Comment (dominiccooney) [inline client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts:23]: Naming, is failing certificate checks "authorization"?

File: client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts [hunk 3]
```diff
@@ -5,9 +5,58 @@ import { isError } from '../../utils'
...
+                            throw new TypeError('expected chunk to be a Buffer')
+                        }
+                        buffer += chunk.toString()
+                    })
+                    res.on('end', () => {
+                        req.destroy()
+                        try {
+                            const resp = JSON.parse(buffer) as CodeCompletionResponse
+                            if (typeof resp.completion !== 'string' || typeof resp.stopReason !== 'string') {
+                                reject(`response ${resp} does not satisfiy CodeCompletionResponse`)
```
Comment (dominiccooney) [inline client/cody-shared/src/sourcegraph-api/completions/nodeClient.ts:39]:
```suggestion
                                reject(`response ${resp} does not satisfy CodeCompletionResponse`)
```

**Review by dominiccooney — APPROVED** [review]

File: client/cody-shared/src/sourcegraph-api/completions/browserClient.ts [hunk 4]
```diff
@@ -1,9 +1,13 @@
 import { fetchEventSource } from '@microsoft/fetch-event-source'
 
 import { SourcegraphCompletionsClient } from './client'
-import type { Event, CompletionParameters, CompletionCallbacks } from './types'
+import type { Event, CompletionParameters, CompletionCallbacks, CodeCompletionResponse } from './types'
 
 export class SourcegraphBrowserCompletionsClient extends SourcegraphCompletionsClient {
+    public complete(): Promise<CodeCompletionResponse> {
+        throw new Error('SourcegraphBrowserCompletionsClienet.complete not implemented')
```
Comment (eseliger) [inline client/cody-shared/src/sourcegraph-api/completions/browserClient.ts:8]:
```suggestion
        throw new Error('SourcegraphBrowserCompletionsClient.complete not implemented')
```

**Review by eseliger — APPROVED** [review]

**Pull request merged**